### ISSUE TITLE
random.set_state using a random.RandomState doesn't properly work when shuffling arrays

### ISSUE THREAD

**Issue by ViaFerrata** [numpy/numpy#9981]
Hello there,

I'd like to shuffle multiple arrays in unison (1.) and then I'd like to do that multiple times with different seeds (2.). However, it looks like `random.set_state` doesn't properly accept a `random.RandomState`.

Here's a short example:
```
a = np.array([1,2,3,4,5])
b = np.array([10,20,30,40,50])

def shuffle_in_unison(a, b):
    r = np.random.RandomState() # different state from /dev/urandom for each call
    state = r.get_state()
    np.random.shuffle(a) # result: array([4, 2, 1, 5, 3])
    np.random.set_state(state)
    np.random.shuffle(b) # result: array([40, 20, 50, 10, 30])
    # -> doesn't work
    return a,b

for i in xrange(10):
    a,b = shuffle_in_unison(a,b) # shuffle a,b and then shuffle them again with a different seed (RandomState) for random.shuffle
    print a,b
```

Is this a bug, or are you not supposed to set the seed for `random.shuffle` in this way?

I'm using numpy v1.13.3 with Python 2.7.13.

Thanks a lot! :)

**Comment by njsmith** [comment]
In your example code, it looks like the first shuffle is using whatever
arbitrary state the global np.random generator happens to have, and only
the second one is using the state you specified? It would probably be
better to avoid using the global generator we entirely (only call methods
on a RandomState object, never global functions in the np.random.*
namespace). But if you do want to use the global namespace like this, I
think you need to call set_state twice.

**Comment by ViaFerrata** [comment]
@njsmith 
Yes, I don't care about the state in the first shuffle. However, the second shuffle needs to reuse the seed from the first shuffle, such that the rows of a and b still correspond to each other.

<hr>

What I want to do is as follows:
1) Call the function: shuffle the arrays `a,b` in unison, such that row `i` of `a` still corresponds to row `i` of `b` . For this, I need the same seed for both shuffles.
2) Do this n times, but use a different seed for each of the n unison shuffles. Otherwise it's probably not completely random if the seed is the same for each iteration (i.e., the `random.shuffle` might move row i to row j for **every** iteration). 

For 1) it's perfectly fine to use the following:
```
def shuffle_in_unison(a, b):
    rng_state = numpy.random.get_state()
    numpy.random.shuffle(a)
    numpy.random.set_state(rng_state)
    numpy.random.shuffle(b)
```

Unfortunately, it doesn't work for iterating, since the state `rng_state = numpy.random.get_state()` is same for each call. That's why I wanted to extend the function in this way with the RandomGenerator.

<hr>

I'm not set on using the global generator, I'd just like to achieve this functionality.
But if I understand you correctly, I'd have to set the state before the first shuffle as well (contrary to the code snippet above in this post)?

**Comment by seberg** [comment]
get_state will not be the same for each call, since the second call also changes it (only after the first one you reset), but it still is probably nicer with RandomState.

**Comment by rkern** [comment]
Never use `np.random.set_state()`, please. Leave that poor hidden global state alone! See this comment for an approach that ought to work for this use case:

https://github.com/numpy/numpy/pull/9906#discussion_r146428000

**Comment by rkern** [comment]
Or if you want an arbitrary starting point instead of a fixed seed:

```
import copy

prng = np.random.RandomState()  # arbitrary state
prng2 = copy.copy(prng)
prng.shuffle(a)
prng2.shuffle(b)
```
